### DISCUSSION TITLE
using step to generate certificates for A.D. with a side question about CSR's

### DISCUSSION THREAD

**Discussion by alsoeric** [smallstep/certificates#1026] [Q&A]
Step is great. Using it to generate certificates for our internal use but the one place and running to troubles with certificates for active directory. It seems to want a certificate based off of a CSR that it generates. 

1. How does when you step to generate certificates for active directory?
2. How does one use a CSR with step?

**Comment by tashian** [comment]
1. By default, step-ca issues TLS certificates. You'll need to create certificate templates that will give you a certificate format compatible with AD. You can find some information about [server](https://docs.microsoft.com/en-us/windows-server/networking/core-network-guide/cncg/server-certs/configure-the-server-certificate-template) and [workstation](https://docs.microsoft.com/en-us/windows/security/threat-protection/windows-firewall/configure-the-workstation-authentication-certificate-template) certificate formats in Microsoft's documentation. But you'll need to translate those requirements into `step-ca` templates. See our [template documentation](https://smallstep.com/docs/step-ca/templates). (And, when you get it working, please publish your templates here and I'm sure anyone else looking to do this would be grateful!)
2. Once you have your templates configured, you'll use the `step ca sign` subcomand to sign a CSR

Hope this helps. :D

**Reply by tashian** [reply]
I also want to add a huge warning to this note. We've chosen sane defaults for TLS certificates in `step-ca`, and we've tried to make it "hard to misuse" or to configure it in an insecure way. When you start writing your own templates, it's easy to unwittingly introduce security vulnerabilities. See [this article](https://posts.specterops.io/certified-pre-owned-d95910965cd2), for example, about how misconfigured AD certificate templates can open up an AD network to attackers.
